### ISSUE TITLE
[Bug] Functional subset for delayed `variable_choices` isn't working for `tm_g_gh_boxplot`

### ISSUE THREAD

**Issue by donyunardi** [insightsengineering/teal.goshawk#301]
## Summary

Per documentation, when operating with delayed data, user can create delayed `variable_choices` using functional subset technique:

https://github.com/insightsengineering/teal.transform/blob/016799c641e9079a4893a95d825d803cd82f6dda/R/choices_labeled.R#L168-L172

```r
#' # functional subset (with delayed data) - return only factor variables
#' variable_choices("ADRS", subset = function(data) {
#'   idx <- vapply(data, is.factor, logical(1))
#'   names(data)[idx]
#' })
```

However, this doesn't seem to be working with `tm_g_gh_boxplot`.
![image](https://github.com/user-attachments/assets/a3497b6a-5e3c-4b70-99d7-f03e3498cf5d)

During initial investigation, the issue is because the delayed `variable_choices` was not resolved when executing teal.widgets::optionalSelectInput:

https://github.com/insightsengineering/teal.goshawk/blob/49175545f79752d786daef87f446f30234b6d503/R/tm_g_gh_boxplot.R#L311-L317

https://github.com/insightsengineering/teal.widgets/blob/2f43fb5fb5bab6974cae4ccc63501c7f74bbd5cf/R/optionalInput.R#L86-L91

<img src="https://github.com/user-attachments/assets/500f72df-3c3e-4684-82c8-c0e6c1439c75" width="50%" height="50%" />

We need investigate further on why the functional subset is not being resolved. 
https://github.com/insightsengineering/teal.transform/blob/a0db9fc819a36f9d181d8829f16af81a3c41d721/R/choices_labeled.R#L204-L206

It could also be related with `teal.goshawk::get_choices` function.

<details>

<summary> Example Code </summary>

```r
data <- teal_data_module(
  ui <- function(id) {
    ns <- NS(id)
    actionButton(ns("submit"), "Submit")
  },
  server = function(id) {
    moduleServer(id, function(input, output, session) {
      eventReactive(input$submit, {
        data <- within(
          teal_data(),
          {
            ADLB <- teal.data::rADLB
            ADSL <- teal.data::rADSL

            library(dplyr)
            library(nestcolor)
            library(stringr)
          #'
            # use non-exported function from goshawk
            h_identify_loq_values <- getFromNamespace("h_identify_loq_values", "goshawk")
          #'
            # original ARM value = dose value
            arm_mapping <- list(
              "A: Drug X" = "150mg QD",
              "B: Placebo" = "Placebo",
              "C: Combination" = "Combination"
            )
            set.seed(1)
            ADSL <- rADSL
            ADLB <- rADLB
            var_labels <- lapply(ADLB, function(x) attributes(x)$label)
            ADLB <- ADLB %>%
              mutate(
                AVISITCD = case_when(
                  AVISIT == "SCREENING" ~ "SCR",
                  AVISIT == "BASELINE" ~ "BL",
                  grepl("WEEK", AVISIT) ~ paste("W", str_extract(AVISIT, "(?<=(WEEK ))[0-9]+")),
                  TRUE ~ as.character(NA)
                ),
                AVISITCDN = case_when(
                  AVISITCD == "SCR" ~ -2,
                  AVISITCD == "BL" ~ 0,
                  grepl("W", AVISITCD) ~ as.numeric(gsub("[^0-9]*", "", AVISITCD)),
                  TRUE ~ as.numeric(NA)
                ),
                AVISITCD = factor(AVISITCD) %>% reorder(AVISITCDN),
                TRTORD = case_when(
                  ARMCD == "ARM C" ~ 1,
                  ARMCD == "ARM B" ~ 2,
                  ARMCD == "ARM A" ~ 3
                ),
                ARM = as.character(arm_mapping[match(ARM, names(arm_mapping))]),
                ARM = factor(ARM) %>% reorder(TRTORD),
                ACTARM = as.character(arm_mapping[match(ACTARM, names(arm_mapping))]),
                ACTARM = factor(ACTARM) %>% reorder(TRTORD),
                ANRLO = 50,
                ANRHI = 75
              ) %>%
              rowwise() %>%
              group_by(PARAMCD) %>%
              mutate(LBSTRESC = ifelse(
                USUBJID %in% sample(USUBJID, 1, replace = TRUE),
                paste("<", round(runif(1, min = 25, max = 30))), LBSTRESC
              )) %>%
              mutate(LBSTRESC = ifelse(
                USUBJID %in% sample(USUBJID, 1, replace = TRUE),
                paste(">", round(runif(1, min = 70, max = 75))), LBSTRESC
              )) %>%
              ungroup()
          #'
            attr(ADLB[["ARM"]], "label") <- var_labels[["ARM"]]
            attr(ADLB[["ACTARM"]], "label") <- var_labels[["ACTARM"]]
            attr(ADLB[["ANRLO"]], "label") <- "Analysis Normal Range Lower Limit"
            attr(ADLB[["ANRHI"]], "label") <- "Analysis Normal Range Upper Limit"
          #'
            # add LLOQ and ULOQ variables
            ALB_LOQS <- h_identify_loq_values(ADLB, "LOQFL")
            ADLB <- left_join(ADLB, ALB_LOQS, by = "PARAM")

          }
        )
        datanames <- c("ADLB", "ADSL")

        datanames(data) <- datanames
        join_keys(data) <- default_cdisc_join_keys[datanames]
        data
      })
    })
  }
)

# functional subset
cs_facet_var <- choices_selected(
  choices = variable_choices("ADLB", function(data) {
    c("ARM", "AVISITCD", names(data)[26:193])
  }),
  selected = "AVISITCD"
)

app <- init(
  data = data,
  modules = modules(
    tm_g_gh_boxplot(
      label = "Box Plot",
      dataname = "ADLB",
      param_var = "PARAMCD",
      param = choices_selected(c("ALT", "CRP", "IGA"), "ALT"),
      yaxis_var = choices_selected(c("AVAL", "BASE", "CHG"), "AVAL"),
      xaxis_var = choices_selected(c("ACTARM", "ARM", "AVISITCD", "STUDYID"), "ARM"),
      facet_var = cs_facet_var,
      trt_group = choices_selected(c("ARM", "ACTARM"), "ARM"),
      loq_legend = TRUE,
      rotate_xlab = FALSE,
      hline_arb = c(60, 55),
      hline_arb_color = c("grey", "red"),
      hline_arb_label = c("default_hori_A", "default_hori_B"),
      hline_vars = c("ANRHI", "ANRLO", "ULOQN", "LLOQN"),
      hline_vars_colors = c("pink", "brown", "purple", "black"),
    )
  ),
  title = "my teal app"
)

shinyApp(app$ui, app$server)

```

</details>

## Acceptance Criteria
* User can perform functional subset when creating delayed `variable_choices`/`choices_selected` when using `tm_g_gh_boxplot`
* Investigate and assess if the solution is also applicable with other `teal.goshawk` modules. 
* If not, investigate, assess, and create actionable issues.

**Comment by npaszty** [comment]
@donyunardi 

thanks for writing the acceptance criteria. I'm assuming that with the fix any module would be able to pass a function?

tm_g_gh_lineplot also uses the same concept of spliting the lines so that's another test case.

**Comment by donyunardi** [comment]
I added a new acceptance criteria to check solution with other teal.goshawk modules.

**Comment by m7pr** [comment]
Hey @donyunardi 

Based on the code provided in the issue I understand that there is a need to create `variable_choices` (used in `choices_selected`) that use a `subset` parameter that is a function that works on a `data` - assuming it's the `input` data to `teal::init`. 

I personally can not find any working example of `variable_choices` with `subset = function(data)` nor in vignettes and nor in `teal.gallery` so I don't really understand how this supposed to work. 

I guess the `functional form of variable_choices for delayed_data` is a coincidence of names with `teal_data_module` and a coincidence that `teal::init` takes `data` and `variable_choices(subset = function(` takes data.

I don't see how `variable_choices` could use `data` as `data` is a part of the server and `variables_choices` is passed to `ui_g_boxplot` that passes them to `teal.widgets::optionalSelectInput` that uses `shinyWidgets::pickerInput`. The last one throws an error that it can not use a function as an input. 

In my opinion the example listed here https://github.com/insightsengineering/teal.transform/blob/016799c641e9079a4893a95d825d803cd82f6dda/R/choices_labeled.R#L168-L172 is wrong or/and outdated. I think it should not contain a function, but an actual realization of the function call. And the function should be realized before it's inputed into `teal::init` so I guess this should happen before teal starts the app. 

The trick in here https://github.com/insightsengineering/teal.modules.clinical/issues/1221 is a hack to assign a variable in the global environment from the `teal_data`. So once `teal_data` is created before the `teal::init` is even called, we already have a variable in the global environment and it can be used in `variables_choices`.

I'm a bit confused and I don't think I will be able to solve this

**Comment by m7pr** [comment]
All credits with this investigation goes to the one and only @vedhav !!

# Wrong documentation

Hey guys, so the documentation of `teal.transform` was wrong

Instead of 

```r
#' variable_choices("ADRS", subset = function(data) {
#'   idx <- vapply(data, is.factor, logical(1))
#'   names(data)[idx]
#' })
```

it supposted to be

```r
#' variable_choices(ADRS, subset = function(data) {
#'   idx <- vapply(data, is.factor, logical(1))
#'   names(data)[idx]
#' })

```

so that we apply `variable_choices` to an actual `data.frame`.

I have fixed the example in `teal.transform` https://github.com/insightsengineering/teal.transform/pull/224

# So wrong the usage

So in your app code, instead of the below

```r
cs_facet_var <- choices_selected(
  choices = variable_choices("ADLB", function(data) {
    c("ARM", "AVISITCD", names(data)[26:193])
  }),
  selected = "AVISITCD"
)
```

you should use

```r
cs_facet_var <- choices_selected(
  choices = variable_choices(ADLB, function(data) {
    c("ARM", "AVISITCD", names(data)[26:193])
  }),
  selected = "AVISITCD"
)
```

to get this working.

# Usage

However `cs_facet_var` uses variables from the `.GlobalEnv` so this data can not be specified in `teal_data_module`. 
It rather can be used for `teal_data` input, where you extract ADLB out of `teal_data`.

Please also note `na.omit(unique())` addition in the extraction of names od ADLB.

```r
library(teal.goshawk)

data <- within(
  teal_data(),
  {
    ADLB <- teal.data::rADLB
    ADSL <- teal.data::rADSL
    
    library(dplyr)
    library(nestcolor)
    library(stringr)
    #'
    # use non-exported function from goshawk
    h_identify_loq_values <- getFromNamespace("h_identify_loq_values", "goshawk")
    #'
    # original ARM value = dose value
    arm_mapping <- list(
      "A: Drug X" = "150mg QD",
      "B: Placebo" = "Placebo",
      "C: Combination" = "Combination"
    )
    set.seed(1)
    ADSL <- rADSL
    ADLB <- rADLB
    var_labels <- lapply(ADLB, function(x) attributes(x)$label)
    ADLB <- ADLB %>%
      mutate(
        AVISITCD = case_when(
          AVISIT == "SCREENING" ~ "SCR",
          AVISIT == "BASELINE" ~ "BL",
          grepl("WEEK", AVISIT) ~ paste("W", str_extract(AVISIT, "(?<=(WEEK ))[0-9]+")),
          TRUE ~ as.character(NA)
        ),
        AVISITCDN = case_when(
          AVISITCD == "SCR" ~ -2,
          AVISITCD == "BL" ~ 0,
          grepl("W", AVISITCD) ~ as.numeric(gsub("[^0-9]*", "", AVISITCD)),
          TRUE ~ as.numeric(NA)
        ),
        AVISITCD = factor(AVISITCD) %>% reorder(AVISITCDN),
        TRTORD = case_when(
          ARMCD == "ARM C" ~ 1,
          ARMCD == "ARM B" ~ 2,
          ARMCD == "ARM A" ~ 3
        ),
        ARM = as.character(arm_mapping[match(ARM, names(arm_mapping))]),
        ARM = factor(ARM) %>% reorder(TRTORD),
        ACTARM = as.character(arm_mapping[match(ACTARM, names(arm_mapping))]),
        ACTARM = factor(ACTARM) %>% reorder(TRTORD),
        ANRLO = 50,
        ANRHI = 75
      ) %>%
      rowwise() %>%
      group_by(PARAMCD) %>%
      mutate(LBSTRESC = ifelse(
        USUBJID %in% sample(USUBJID, 1, replace = TRUE),
        paste("<", round(runif(1, min = 25, max = 30))), LBSTRESC
      )) %>%
      mutate(LBSTRESC = ifelse(
        USUBJID %in% sample(USUBJID, 1, replace = TRUE),
        paste(">", round(runif(1, min = 70, max = 75))), LBSTRESC
      )) %>%
      ungroup()
    #'
    attr(ADLB[["ARM"]], "label") <- var_labels[["ARM"]]
    attr(ADLB[["ACTARM"]], "label") <- var_labels[["ACTARM"]]
    attr(ADLB[["ANRLO"]], "label") <- "Analysis Normal Range Lower Limit"
    attr(ADLB[["ANRHI"]], "label") <- "Analysis Normal Range Upper Limit"
    #'
    # add LLOQ and ULOQ variables
    ALB_LOQS <- h_identify_loq_values(ADLB, "LOQFL")
    ADLB <- left_join(ADLB, ALB_LOQS, by = "PARAM")
    
  }
)
datanames <- c("ADLB", "ADSL")

datanames(data) <- datanames
join_keys(data) <- default_cdisc_join_keys[datanames]


# functional subset
cs_facet_var <- choices_selected(
  choices = variable_choices(data[['ADLB']], function(data) {
    na.omit(unique(c("ARM", "AVISITCD", names(data)[26:193])))
  }),
  selected = "AVISITCD"
)

app <- init(
  data = data,
  modules = modules(
    tm_g_gh_boxplot(
      label = "Box Plot",
      dataname = "ADLB",
      param_var = "PARAMCD",
      param = choices_selected(c("ALT", "CRP", "IGA"), "ALT"),
      yaxis_var = choices_selected(c("AVAL", "BASE", "CHG"), "AVAL"),
      xaxis_var = choices_selected(c("ACTARM", "ARM", "AVISITCD", "STUDYID"), "ARM"),
      facet_var = cs_facet_var,
      trt_group = choices_selected(c("ARM", "ACTARM"), "ARM"),
      loq_legend = TRUE,
      rotate_xlab = FALSE,
      hline_arb = c(60, 55),
      hline_arb_color = c("grey", "red"),
      hline_arb_label = c("default_hori_A", "default_hori_B"),
      hline_vars = c("ANRHI", "ANRLO", "ULOQN", "LLOQN"),
      hline_vars_colors = c("pink", "brown", "purple", "black"),
    )
  ),
  title = "my teal app"
)

shinyApp(app$ui, app$server)

```

**Comment by m7pr** [comment]
@donyunardi this is only for `teal_data()` for `teal_data_module` I think it's not possible to use `variables_choices` for this one, but you can always `updatePickerInput(` of the variable in the server part of `teal_data_module`. I will prepare an example

**Comment by m7pr** [comment]
I guess we could use `teal_data_module` with some random `teal_transform_module` that updated the names, once it's UI button is clicked. I have a prototype below, but it doesn't update the names on the click, so it needs some work.

```r
data <- teal_data_module(
  ui <- function(id) {
    ns <- NS(id)
    actionButton(ns("submit"), "Submit")
  },
  server = function(id) {
    moduleServer(id, function(input, output, session) {
      eventReactive(input$submit, {
        data <- within(
          teal_data(),
          {
            ADLB <- teal.data::rADLB
            ADSL <- teal.data::rADSL
            
            library(dplyr)
            library(nestcolor)
            library(stringr)
            #'
            # use non-exported function from goshawk
            h_identify_loq_values <- getFromNamespace("h_identify_loq_values", "goshawk")
            #'
            # original ARM value = dose value
            arm_mapping <- list(
              "A: Drug X" = "150mg QD",
              "B: Placebo" = "Placebo",
              "C: Combination" = "Combination"
            )
            set.seed(1)
            ADSL <- rADSL
            ADLB <- rADLB
            var_labels <- lapply(ADLB, function(x) attributes(x)$label)
            ADLB <- ADLB %>%
              mutate(
                AVISITCD = case_when(
                  AVISIT == "SCREENING" ~ "SCR",
                  AVISIT == "BASELINE" ~ "BL",
                  grepl("WEEK", AVISIT) ~ paste("W", str_extract(AVISIT, "(?<=(WEEK ))[0-9]+")),
                  TRUE ~ as.character(NA)
                ),
                AVISITCDN = case_when(
                  AVISITCD == "SCR" ~ -2,
                  AVISITCD == "BL" ~ 0,
                  grepl("W", AVISITCD) ~ as.numeric(gsub("[^0-9]*", "", AVISITCD)),
                  TRUE ~ as.numeric(NA)
                ),
                AVISITCD = factor(AVISITCD) %>% reorder(AVISITCDN),
                TRTORD = case_when(
                  ARMCD == "ARM C" ~ 1,
                  ARMCD == "ARM B" ~ 2,
                  ARMCD == "ARM A" ~ 3
                ),
                ARM = as.character(arm_mapping[match(ARM, names(arm_mapping))]),
                ARM = factor(ARM) %>% reorder(TRTORD),
                ACTARM = as.character(arm_mapping[match(ACTARM, names(arm_mapping))]),
                ACTARM = factor(ACTARM) %>% reorder(TRTORD),
                ANRLO = 50,
                ANRHI = 75
              ) %>%
              rowwise() %>%
              group_by(PARAMCD) %>%
              mutate(LBSTRESC = ifelse(
                USUBJID %in% sample(USUBJID, 1, replace = TRUE),
                paste("<", round(runif(1, min = 25, max = 30))), LBSTRESC
              )) %>%
              mutate(LBSTRESC = ifelse(
                USUBJID %in% sample(USUBJID, 1, replace = TRUE),
                paste(">", round(runif(1, min = 70, max = 75))), LBSTRESC
              )) %>%
              ungroup()
            #'
            attr(ADLB[["ARM"]], "label") <- var_labels[["ARM"]]
            attr(ADLB[["ACTARM"]], "label") <- var_labels[["ACTARM"]]
            attr(ADLB[["ANRLO"]], "label") <- "Analysis Normal Range Lower Limit"
            attr(ADLB[["ANRHI"]], "label") <- "Analysis Normal Range Upper Limit"
            #'
            # add LLOQ and ULOQ variables
            ALB_LOQS <- h_identify_loq_values(ADLB, "LOQFL")
            ADLB <- left_join(ADLB, ALB_LOQS, by = "PARAM")
            
          }
        )
        datanames <- c("ADLB", "ADSL")
        
        datanames(data) <- datanames
        join_keys(data) <- default_cdisc_join_keys[datanames]
        
        data
      })
    })
  }
)

names_transformer <-   teal_transform_module(
  ui <- function(id) {
    ns <- NS(id)
    actionButton(ns("submit"), "Update facet_var names")
  },
  server = function(id, data) {
    moduleServer(id, function(input, output, session) {
      
      observeEvent(input$submit, {
        names_for_the_update <- na.omit(unique(c(c("ARM", "ACTARM"), names(data[['ADLB']])[26:100])))

        shinyWidgets::updatePickerInput(
          session = session,
          inputId = "teal-teal_modules-box_plot-module-facet_var_input",
          choices = names_for_the_update
        )
      })
      reactive(data())
      
    })
  }
)

boxplot_with_transformers <-
  tm_g_gh_boxplot(
    label = "Box Plot",
    dataname = "ADLB",
    param_var = "PARAMCD",
    param = choices_selected(c("ALT", "CRP", "IGA"), "ALT"),
    yaxis_var = choices_selected(c("AVAL", "BASE", "CHG"), "AVAL"),
    xaxis_var = choices_selected(c("ACTARM", "ARM", "AVISITCD", "STUDYID"), "ARM"),
    facet_var = choices_selected(c("ARM", "AVISITCD"), "AVISITCD"),
    trt_group = choices_selected(c("ARM", "ACTARM"), "ARM"),
    loq_legend = TRUE,
    rotate_xlab = FALSE,
    hline_arb = c(60, 55),
    hline_arb_color = c("grey", "red"),
    hline_arb_label = c("default_hori_A", "default_hori_B"),
    hline_vars = c("ANRHI", "ANRLO", "ULOQN", "LLOQN"),
    hline_vars_colors = c("pink", "brown", "purple", "black")
  )

boxplot_with_transformers$transformers = list(names_transformer)

app <- init(
  data = data,
  modules = modules(
    boxplot_with_transformers
  ),
  title = "my teal app"
)

shinyApp(app$ui, app$server)

```

**Comment by m7pr** [comment]
Or we can use the trick and try to assign a global variable inside teal_data_module server part so that it is used in choices_selected

**Comment by gogonzo** [comment]
What probably should be done is something similar to the fix in [tm_abnormality](https://github.com/insightsengineering/teal.modules.clinical/blob/f2895ee28fd7a136d4fe5151f40b9f2106bef170/R/tm_t_abnormality.R#L540).
Delayed choices_selected needs to be resolved

**Comment by m7pr** [comment]
Hey, in this package, only the `xaxis_var`, `yaxis_var` and `param` parameters could be passed as `delayed_data`.
`facet_var` parameter in `tm_g_gh_boxplot` could not be passed as `delayed_data` base on functional form of `variables_choices`.

Thanks to the collective ideas from @gogonzo and brainstorms with @vedhav I was able to adjust `tm_g_gh_boxplot` so it's `facet_var` parameter can be passed as `delayed_data`, based on how it was implemented for other parameters.

Here is the PR https://github.com/insightsengineering/teal.goshawk/pull/313

The last question is, if we want to extend this approach to `trt_group` parameter that is one more parameter that can be passed as `variables_choices` but it can not yet be passed as `delayed_data` (yet!). If yes, we can incorporate this change in all modules. `trt_group` exists in all other modules. `facet_var` existed only in `tm_g_gh_boxplot`.

**Comment by m7pr** [comment]
PR for `trt_group` is here https://github.com/insightsengineering/teal.goshawk/pull/314
Let me know @donyunardi if there anything else you need

**Comment by donyunardi** [comment]
> Thanks to the collective ideas from @gogonzo and brainstorms with @vedhav I was able to adjust tm_g_gh_boxplot so it's facet_var parameter can be passed as delayed_data, based on how it was implemented for other parameters.

It's interesting to discover that teal modules need to be written in a specific way to enable the functional subset on delayed data when creating variable_choices. 

From what I've observed, this is not clearly communicated anywhere in our documentation, especially under the topic of creating custom teal modules. We should make this information more visible in our docs or reassess whether this is the right approach for subsetting data or objects from a `teal_data` object.

**Comment by donyunardi** [comment]
> The last question is, if we want to extend this approach to trt_group parameter that is one more parameter that can be passed as variables_choices but it can not yet be passed as delayed_data (yet!). 

Yes, I would think we want to extend this to trt_group parameter.
Can we also check other modules arguments that accepts `choices_selected`? For example, I think `tm_g_gh_scatterplot` needs resolved delated for param, xaxis_var, yaxis_var, and trt_group (which you already handled).

**Comment by donyunardi** [comment]
Can we also add this feature to  `filter_var` in `tm_g_gh_spaghettiplot`?

**Comment by m7pr** [comment]
Hey @donyunardi for `filter_var` parameter, I believe it only exists in the documentation, but is not used in the code anymore https://github.com/search?q=repo%3Ainsightsengineering%2Fteal.goshawk%20filter_var&type=code

**Comment by donyunardi** [comment]
I agree. Also, I don't think we need `filter_var` parameter for `tm_g_gh_spaghettiplot` since it's not being used at all in the code. We should remove this. 

I'll create a separate issue.

Closing this issue as this has been completed.
